lib/schedule/data.ex

```diff
@@ -214,87 +266,182 @@ defmodule Schedule.Data do
-    active_services = Map.take(calendar, dates)
 
-    Enum.flat_map(active_services, fn {_data, service_ids} ->
+    dates
+    |> active_services_on_dates(tables)
+    |> Enum.flat_map(fn {_data, service_ids} ->
       Enum.flat_map(service_ids, fn service_id ->
-        Map.get(swings, {service_id, route_id}, [])
+        case :mnesia.dirty_match_object({swings_table, {service_id, route_id}, :_}) do
+          swing_records when is_list(swing_records) -> Enum.map(swing_records, &elem(&1, 2))
+          _ -> []
+        end
       end)
     end)
   end
 
   # Initialization
 
+  @spec initialize_tables(tables()) :: :ok
+  def initialize_tables(tables) do
+    Enum.each(@table_schema, fn {table_key, columns, extra_indices} ->
+      # TODO: drop table if it already exists before attempting to create
+      {:atomic, :ok} =
+        :mnesia.create_table(tables[table_key],
+          attributes: columns,
+          index: extra_indices,
+          storage_properties: [ets: [{:read_concurrency, true}]],
+          local_content: true
+        )
+
+      :mnesia.wait_for_tables([tables[table_key]], 5_000)
+    end)
+
+    :ok
+  end
+
+  @spec save_schedule_data_to_tables(tables(), t()) :: :ok
+  def save_schedule_data_to_tables(tables, schedule_data) do
+    write_data = fn ->
+      Enum.each(schedule_data.routes, fn route ->
+        :mnesia.write({tables.routes, route.id, route})
+      end)
+
+      Enum.each(schedule_data.route_patterns, fn route_pattern ->
+        :mnesia.write({tables.route_patterns, route_pattern.id, route_pattern})
+      end)
+
+      Enum.each(schedule_data.timepoints_by_route, fn {route_id, timepoints} ->
+        :mnesia.write({tables.timepoints_by_route, route_id, timepoints})
+      end)
+
+      Enum.each(schedule_data.timepoint_names_by_id, fn {id, timepoint_name} ->
+        :mnesia.write({tables.timepoint_names_by_id, id, timepoint_name})
+      end)
+
+      Enum.each(schedule_data.shapes, fn {route_id, shapes} ->
+        :mnesia.write({tables.shapes, route_id, shapes})
+      end)
+
+      Enum.each(schedule_data.stops, fn {id, stop} -> :mnesia.write({tables.stops, id, stop}) end)
+
+      Enum.each(schedule_data.trips, fn {id, trip} ->
+        :mnesia.write({tables.trips, id, trip.service_id, trip})
+      end)
+
+      Enum.each(schedule_data.blocks, fn {id, block} ->
+        :mnesia.write({tables.blocks, id, block.service_id, block})
+      end)
+
+      Enum.each(schedule_data.calendar, fn {date, service_ids} ->
+        :mnesia.write({tables.calendar, date, service_ids})
+      end)
+
+      Enum.each(schedule_data.runs, fn {key, run} ->
+        :mnesia.write({tables.runs, key, run.service_id, run})
+      end)
+
+      Enum.each(schedule_data.swings, fn {{service_id, route_id}, swings} ->
+        :mnesia.write({tables.swings, {service_id, route_id}, swings})
+      end)
+    end
+
+    {:atomic, :ok} = :mnesia.transaction(write_data)
```

Comment:
praise: You know, I never considered writing to all the tables as a single transaction in the API. I suppose it's trickier since each one is currently a separate process, but definitely something to think about in the future!